src/main/java/pl/edu/agh/firecell/engine/algorithm/SmokePropagator.java

```diff
@@ -0,0 +1,129 @@
+                isCellBurning(oldState.getCell(NeighbourUtils.west(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.west(cellIndex)).isSolid()) {
+            westCell = Optional.of(oldState.getCell(NeighbourUtils.west(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.east(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.east(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.east(cellIndex)).isSolid()) {
+            eastCell = Optional.of(oldState.getCell(NeighbourUtils.east(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.north(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.north(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.north(cellIndex)).isSolid()) {
+            northCell = Optional.of(oldState.getCell(NeighbourUtils.north(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.south(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.south(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.south(cellIndex)).isSolid()) {
+            southCell = Optional.of(oldState.getCell(NeighbourUtils.south(cellIndex)));
+        }
+        if (southCell.isPresent()) smokeFromFire += southCell.get().material().smokeCoe() / 4;
+        if (northCell.isPresent()) smokeFromFire += northCell.get().material().smokeCoe() / 4;
+        if (westCell.isPresent()) smokeFromFire += westCell.get().material().smokeCoe() / 4;
+        if (eastCell.isPresent()) smokeFromFire += eastCell.get().material().smokeCoe() / 4;
+
+        double smokeDifference = getSmokeIndicatorDifference(oldCell, oldState, cellIndex);
+        if(Math.min(oldCell.smokeIndicator() + smokeDifference + smokeFromFire, 100) < 0.0005)
+            return 0.0;
+        return Math.min(oldCell.smokeIndicator() + smokeDifference + smokeFromFire, 100);
+    }
+
+    private double getSmokeIndicatorDifference(Cell oldCell, State oldState, Vector3i cellIndex) {
+        Int counter = new Int(oldCell.smokeIndicator());
+        return NeighbourUtils.neighboursStream(cellIndex)
+                .filter(oldState::hasCell)
+                .filter(neighbourIndex -> oldState.getCell(neighbourIndex).isFluid())
+                .map(neighbourIndex -> {
+                    double diff = 0.0;
+                    if (NeighbourUtils.up(cellIndex).equals(neighbourIndex)) {
+                        diff = -Math.min(oldCell.smokeIndicator(), 100 - oldState.getCell(neighbourIndex).smokeIndicator());
+                        counter.v += diff;
+                    } else if (NeighbourUtils.down(cellIndex).equals(neighbourIndex)) {
+                        diff = Math.min(oldState.getCell(neighbourIndex).smokeIndicator(),
+                                100 - oldCell.smokeIndicator());
+                        counter.v += diff;
+                    } else if (isHorizontal(cellIndex, neighbourIndex)) {
```

Comment:
... since it is the last else-if branch, we no longer need this else if here (because we checked for up and down neighbour, now it has to be horizointal one  ...